Games/Chess/Chess.cpp

```diff
@@ -0,0 +1,379 @@
+        int rank_delta = abs(move.to.rank - move.from.rank);
+        int file_delta = abs(move.to.file - move.from.file);
+        if (get_piece(move.to).colour != colour && max(rank_delta, file_delta) == 2 && min(rank_delta, file_delta) == 1) {
+            return true;
+        }
+    } else if (piece.type == Type::Bishop) {
+        int rank_delta = move.to.rank - move.from.rank;
+        int file_delta = move.to.file - move.from.file;
+        if (abs(rank_delta) == abs(file_delta)) {
+            int dr = rank_delta / abs(rank_delta);
+            int df = file_delta / abs(file_delta);
+            for (Square sq = move.from; sq != move.to; sq.rank += dr, sq.file += df) {
+                if (get_piece(sq).type != Type::None && sq != move.from) {
+                    return false;
+                }
+            }
+
+            if (get_piece(move.to).colour != colour) {
+                return true;
+            }
+        }
+    } else if (piece.type == Type::Rook) {
+        int rank_delta = move.to.rank - move.from.rank;
+        int file_delta = move.to.file - move.from.file;
+        if (rank_delta == 0 || file_delta == 0) {
+            int dr = (rank_delta) ? rank_delta / abs(rank_delta) : 0;
+            int df = (file_delta) ? file_delta / abs(file_delta) : 0;
+            for (Square sq = move.from; sq != move.to; sq.rank += dr, sq.file += df) {
+                if (get_piece(sq).type != Type::None && sq != move.from) {
+                    return false;
+                }
+            }
+
+            if (get_piece(move.to).colour != colour) {
+                return true;
+            }
+        }
+    } else if (piece.type == Type::Queen) {
+        int rank_delta = move.to.rank - move.from.rank;
+        int file_delta = move.to.file - move.from.file;
+        if (abs(rank_delta) == abs(file_delta) || rank_delta == 0 || file_delta == 0) {
+            int dr = (rank_delta) ? rank_delta / abs(rank_delta) : 0;
+            int df = (file_delta) ? file_delta / abs(file_delta) : 0;
+            for (Square sq = move.from; sq != move.to; sq.rank += dr, sq.file += df) {
+                if (get_piece(sq).type != Type::None && sq != move.from) {
+                    return false;
+                }
+            }
+
+            if (get_piece(move.to).colour != colour) {
+                return true;
+            }
+        }
+    } else if (piece.type == Type::King) {
+        int rank_delta = move.to.rank - move.from.rank;
+        int file_delta = move.to.file - move.from.file;
+        if (abs(rank_delta) <= 1 && abs(file_delta) <= 1) {
+            if (get_piece(move.to).colour != colour) {
+                return true;
+            }
+        }
+
+        if (colour == Colour::White) {
+            if ((move.to == Square("a1") || move.to == Square("c1")) && m_white_can_castle_queenside && get_piece(Square("b1")).type == Type::None && get_piece(Square("c1")).type == Type::None && get_piece(Square("d1")).type == Type::None) {
```

Comment:
You need to check that the squares the king moves over are not targeted by an enemy piece too :)

Reply:
nice catch :^). this is now implemented in `is_legal` rather than `is_legal_no_check` to prevent infinite recursion.